contracts/contracts/vault/VaultCore.sol

```diff
@@ -385,32 +385,102 @@ contract VaultCore is VaultInitializer {
      * @return totalUnits Total balance of Vault in units
      */
     function _rebase() internal whenNotRebasePaused returns (uint256) {
-        uint256 ousdSupply = oUSD.totalSupply();
+        uint256 supply = oUSD.totalSupply();
         uint256 vaultValue = _totalValue();
-        if (ousdSupply == 0) {
+        uint256 nonRebasing = oUSD.nonRebasingSupply();
+        // If no supply yet, do not rebase
+        if (supply == 0) {
             return vaultValue;
         }
 
-        // Yield fee collection
+        // Calculate yield and new supply
+        uint256 yield = 0;
+        uint256 newTarget = 0;
+        (yield, newTarget) = _nextYield(supply, vaultValue, nonRebasing);
+        uint256 newSupply = supply + yield;
+        // Only rebase upwards and if we have enough backing funds
+        if (newSupply <= supply || newSupply > vaultValue) {
+            return vaultValue;
+        }
+
+        // Fee collection on yield
         address _trusteeAddress = trusteeAddress; // gas savings
-        if (_trusteeAddress != address(0) && (vaultValue > ousdSupply)) {
-            uint256 yield = vaultValue - ousdSupply;
-            uint256 fee = yield.mulTruncateScale(trusteeFeeBps, 1e4);
-            require(yield > fee, "Fee must not be greater than yield");
+        uint256 fee = 0;
+        if (_trusteeAddress != address(0)) {
+            fee = (yield * trusteeFeeBps) / 1e4;
             if (fee > 0) {
+                require(fee < yield, "Fee must not be greater than yield");
                 oUSD.mint(_trusteeAddress, fee);
             }
-            emit YieldDistribution(_trusteeAddress, yield, fee);
         }
+        emit YieldDistribution(_trusteeAddress, yield, fee);
 
         // Only ratchet OToken supply upwards
-        ousdSupply = oUSD.totalSupply(); // Final check should use latest value
-        if (vaultValue > ousdSupply) {
-            oUSD.changeSupply(vaultValue);
+        // Final check uses latest totalSupply
+        uint256 preSupply = oUSD.totalSupply();
+        if (newSupply > preSupply) {
+            oUSD.changeSupply(newSupply);
+            rebasePerSecondTarget = uint64(_min(newTarget, type(uint64).max));
+            lastRebase = uint64(block.timestamp); // Intentional cast
         }
         return vaultValue;
     }
 
+    /**
+     * @notice Calculates the amount that would rebase at at next rebase.
+     * This is before any fees.
+     * @return yield amount of expected yield
+     */
+    function previewYield() external view returns (uint256 yield) {
+        uint256 _totalSupply = oUSD.totalSupply();
+        uint256 _nonRebasing = oUSD.nonRebasingSupply();
```

Comment:
nit: both these don't need variables and can be directly called in `_nextYield` function call below